src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -266,138 +343,181 @@ impl WorkunitStore {
+  }
+
+  fn add_completed_workunit_to_store(
     span_id: SpanId,
     new_metadata: Option<WorkunitMetadata>,
-  ) -> Result<(), String> {
+    end_time: SystemTime,
+    inner_store: &mut WorkUnitInnerStore,
+  ) {
     use std::collections::hash_map::Entry;
-    let inner = &mut self.inner.lock();
-    match inner.workunit_records.entry(span_id.clone()) {
-      Entry::Vacant(_) => Err(format!(
-        "No previously-started workunit found for id: {}",
-        span_id
-      )),
+
+    match inner_store.workunit_records.entry(span_id.clone()) {
+      Entry::Vacant(_) => {
+        log::warn!("No previously-started workunit found for id: {}", span_id);
+      }
       Entry::Occupied(o) => {
         let (span_id, mut workunit) = o.remove_entry();
         let time_span = match workunit.state {
           WorkunitState::Completed { .. } => {
-            return Err(format!("Workunit {} was already completed", span_id))
+            log::warn!("Workunit {} was already completed", span_id);
+            return;
+          }
+          WorkunitState::Started { start_time } => {
+            TimeSpan::from_start_and_end_systemtime(&start_time, &end_time)
           }
-          WorkunitState::Started { start_time } => TimeSpan::since(&start_time),
         };
         let new_state = WorkunitState::Completed { time_span };
         workunit.state = new_state;
         if let Some(metadata) = new_metadata {
           workunit.metadata = metadata;
         }
         workunit.log_workunit_state();
-        inner.workunit_records.insert(span_id.clone(), workunit);
-        inner.completed_ids.push(span_id);
-        Ok(())
+        inner_store.workunit_records.insert(span_id, workunit);
       }
     }
   }
 
   pub fn add_completed_workunit(
     &self,
     name: String,
-    time_span: TimeSpan,
+    start_time: SystemTime,
+    end_time: SystemTime,
     parent_id: Option<SpanId>,
     metadata: WorkunitMetadata,
   ) {
-    let inner = &mut self.inner.lock();
     let span_id = new_span_id();
+
     let workunit = Workunit {
       name,
       span_id: span_id.clone(),
       parent_id,
-      state: WorkunitState::Completed { time_span },
+      state: WorkunitState::Started { start_time },
       metadata,
     };
 
-    inner.workunit_records.insert(span_id.clone(), workunit);
-    inner.completed_ids.push(span_id);
+    let sender = self.heavy_hitters_data.started_workunits_tx.lock();
+    sender.send(workunit.clone()).unwrap();
+    let sender = self.streaming_workunit_data.started_workunits_tx.lock();
+    sender.send(workunit).unwrap();
+
+    let sender = self.heavy_hitters_data.completed_workunits_tx.lock();
+    sender.send((span_id.clone(), None, end_time)).unwrap();
+    let sender = self.streaming_workunit_data.completed_workunits_tx.lock();
+    sender.send((span_id, None, end_time)).unwrap();
   }
 
   pub fn with_latest_workunits<F, T>(&mut self, max_verbosity: log::Level, f: F) -> T
```

Comment:
This method is only called by the Streaming workunit handler, right? Could maybe move there?